### PR TITLE
Fix local static variables detection in assigns clauses

### PR DESCRIPTION
**Pull request by remi-delmas-3000**
Function contracts instrumentation:

1. Use the call graph of the function to detect local static variables declared in the functions it calls.

Other changes:
1. subfunction call inlining was moved from `code_contractst::check_frame_conditions` to `code_contractst::check_frame_conditions_function`.
2. Recursive functions are now detected during inlining and cause an INVARIANT violation
3. Function pointers are now detected during call graph construction and cause a PRECONDITION violation

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-instrument/contracts/contracts.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/contracts.cpp]: Wouldn't this create CARs for ALL static local variables, like beyond the current function on which contracts are checked and its subfunction calls?

It might have a significant perf impact if lots of unrelated functions also introduce static local variables, because we'll add their CARs unnecessarily to the write set and check every assignment against them.

Comment (SaswatPadhi) [reply src/goto-instrument/contracts/contracts.cpp]: It might be worth fixing the inlining routine instead. Inlined static locals should just behave as static locals within the caller. It seems weird and inconsistent to have the function body inlined into a caller but be mutating a static local that's tagged with a different function name.

Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/contracts.cpp]: Indeed. 
Since all sub-function calls are inlined for contract checking, a better solution would be to detect static locals used in the body of the function under analysis rather than at the symbol table level.

**Comment by remi-delmas-3000** [comment]
Updated to a better solution is to detect static locals used in the body of the function under analysis rather than at the symbol table level.

File: regression/contracts/assigns_enforce_statics_subfunctions/main.c
Comment (feliperodri) [inline regression/contracts/assigns_enforce_statics_subfunctions/main.c]: You don't need to include this include here.

File: regression/contracts/assigns_enforce_statics_subfunctions/test.desc
Comment (feliperodri) [inline regression/contracts/assigns_enforce_statics_subfunctions/test.desc]: nit:
```Makefile
in assigns clause verification, across sub-function calls.
```

File: regression/contracts/assigns_enforce_statics_subfunctions/test.desc
Comment (feliperodri) [inline regression/contracts/assigns_enforce_statics_subfunctions/test.desc]: Why do we need `--pointer-primitive-check` here?

File: src/goto-instrument/contracts/contracts.cpp
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.cpp]: nit: we could call it `static_local_set` instead of `tmp_set`.

File: src/goto-instrument/contracts/contracts.h
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.h]:
```suggestion
  /// \brief Finds static local symbols that occur in the body of the given function and adds them to the the write set of `assigns`.
```

File: src/goto-instrument/contracts/contracts.h
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.h]:
```suggestion
  /// @param[in]  function The function to search local static symbols in
  /// @param[out] assigns  Assigns clause where search results are added
```

File: src/goto-instrument/contracts/contracts.h
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.h]:
```suggestion
  /// A symbol is considered a static local symbol iff:
  /// - it has a static lifetime annotation;
  /// - its source location has a non-empty function attribute.
```

File: src/goto-instrument/contracts/contracts.h
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.h]:
```suggestion
  /// Symbol occurrences are searched in:
  /// - ASSERT/ASSUME conditions;
  /// - ASSIGN lhs and rhs;
  /// - FUNCTION_CALL function, lhs and operands;
  /// - SET_RETURN_VALUE instructions;
  /// - OTHER code.
```

File: src/goto-instrument/contracts/contracts.h
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.h]:
```suggestion
  /// The search does *not*:
  /// - recurse into FUNCTION_CALL;
  /// - test DECL or DEAD symbols (because local statics are never DECL'd or DEAD'd for being static);
  /// - test attributes of GOTO, THREAD*, ATOMIC*, THROW, CATCH  instructions.
```

File: src/goto-instrument/contracts/contracts.cpp
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.cpp]:
```suggestion
    if(sym.is_static_lifetime && !sym.location.get_function().empty())
```

File: src/goto-instrument/contracts/contracts.h
Comment (feliperodri) [reply src/goto-instrument/contracts/contracts.h]: This function doesn't have a return value.

File: src/goto-instrument/contracts/contracts.cpp
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.cpp]: Could we add more tests to cover all these cases?

**Review by feliperodri — CHANGES_REQUESTED** [review]
Only minor requests.

File: src/goto-instrument/contracts/contracts.cpp
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.cpp]: @tautschnig is there anything already available in CBMC that could help us to get all static local symbols of a given function?

Comment (SaswatPadhi) [reply src/goto-instrument/contracts/contracts.cpp]: I was thinking of something simpler like:

```c++
// record all functions directly/indirectly called by foo
<some_data_structure> foo_calls = call_tree(foo);

// check if a global static variable comes from one of these functions
for(const auto &sym_pair : symbol_table)
  if(sym.is_static_lifetime && foo_calls.contains(sym.location.get_function()))
    local_statics.add(sym)
```

to find all local `static`s indirectly referred to by `foo`.

Do we have a CBMC library function that returns a call tree below a given function `foo`?

File: regression/contracts/assigns_enforce_statics_subfunctions/test.desc
Comment (remi-delmas-3000) [reply regression/contracts/assigns_enforce_statics_subfunctions/test.desc]: I think it's safe to use subfunction
https://www.merriam-webster.com/dictionary/subroutine

Comment (feliperodri) [reply regression/contracts/assigns_enforce_statics_subfunctions/test.desc]: It's fine to use subroutine, the core issue was the typo: `s/accross/across/g`.

File: src/goto-instrument/contracts/contracts.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/contracts.cpp]: the contents of this set only gets filtered for local statics in a second step

File: regression/contracts/assigns_enforce_statics/main.c
Comment (feliperodri) [inline regression/contracts/assigns_enforce_statics/main.c]: We don't need these includes here.

File: regression/contracts/assigns_enforce_statics_subfunctions/test1.c
Comment (feliperodri) [inline regression/contracts/assigns_enforce_statics_subfunctions/test1.c]: We don't need these include here.

File: regression/contracts/assigns_enforce_statics_subfunctions/test2.c
Comment (feliperodri) [inline regression/contracts/assigns_enforce_statics_subfunctions/test2.c]: We don't need this include here.

File: regression/contracts/assigns_enforce_statics_subfunctions/test2.desc
Comment (feliperodri) [inline regression/contracts/assigns_enforce_statics_subfunctions/test2.desc]: Are we tracking this request on a GitHub issue? We should add the link here if yes.

**Review by feliperodri — CHANGES_REQUESTED** [review]
Almost there.

File: regression/contracts/assigns_enforce_statics/main.c
Comment (remi-delmas-3000) [reply regression/contracts/assigns_enforce_statics/main.c]: fixed

File: src/goto-instrument/contracts/contracts.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/contracts.cpp]: now using the call_graph of the function under analysis to perform the detection

File: regression/contracts/assigns_enforce_statics_subfunctions/main.c
Comment (feliperodri) [inline regression/contracts/assigns_enforce_statics_subfunctions/main.c]: We still don't need these here.

File: regression/contracts/assigns_enforce_statics_subfunctions/test.desc
Comment (feliperodri) [reply regression/contracts/assigns_enforce_statics_subfunctions/test.desc]: ?

File: src/goto-instrument/contracts/contracts.cpp
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.cpp]: Why don't we hit this case?

File: src/goto-instrument/contracts/contracts.h
Comment (feliperodri) [inline src/goto-instrument/contracts/contracts.h]:
```suggestion
  /// \brief Finds symbols declared with a static lifetime in the given
  /// `root_function` or one of the functions it calls,
  /// and adds them to the given `assigns`.
  ///
  /// @param functions     All functions of the `goto_model`.
  /// @param root_function The root function under which to search statics.
  /// @param assigns       Assigns clause where search results are added.
  ///
  /// A symbol is considered a static local symbol iff:
  /// - it has a static lifetime annotation;
  /// - its source location has a non-empty function attribute;
  /// - this function attribute is found in the call graph of the `root_function`.
```

**Review by feliperodri — APPROVED** [review]
LGTM modulo a few minor comments.

**Comment by remi-delmas-3000** [comment]
We are now collecting symbols from the symbol table that have a static lifetime and are declare in functions occurring in the  call graph of the function under analysis

File: regression/contracts/assigns_enforce_statics_subfunctions/test.desc
Comment (tautschnig) [reply regression/contracts/assigns_enforce_statics_subfunctions/test.desc]: Why is there an underscore here (after `bar`)?

File: regression/contracts/assigns_enforce_statics/main.c
Comment (tautschnig) [inline regression/contracts/assigns_enforce_statics/main.c:1]: Random place for comment: the commit message ...
1) Could make clear that this is about code contracts;
2) Should likely have the "A symbol is ..." all moved into the body of the message;
3) Should likely talk about "called functions" instead of "subfunctions."

File: src/goto-instrument/contracts/contracts.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/contracts.cpp]: Nit picking: please move the `util` includes a few lines down, where the other `util` includes are.

File: src/goto-instrument/contracts/contracts.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/contracts.cpp]: Much like the commit message: I don't think "subfunctions" is the right terminology.

File: src/goto-instrument/contracts/contracts.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/contracts.cpp]: I think the comment could be more useful if it explained _why_ full inlining was necessary. (Also, will this work with recursive functions?)

File: src/goto-instrument/contracts/contracts.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/contracts.cpp]: What about function pointers?

File: regression/contracts/assigns_enforce_statics_subfunctions/test.desc
Comment (remi-delmas-3000) [reply regression/contracts/assigns_enforce_statics_subfunctions/test.desc]: this underscore separates two groups of options , the first group is passed to goto-instrument to perform the contract instrumentation, and the second set is passed to cbmc to analyse the instrumented contracts.

File: src/goto-instrument/contracts/contracts.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/contracts.cpp]: done

File: src/goto-instrument/contracts/contracts.cpp
Comment (SaswatPadhi) [reply src/goto-instrument/contracts/contracts.cpp]: I'm curious too. Shouldn't we hit this on `regression/contracts/assigns_enforce_statics_subfunctions` test?

Comment (SaswatPadhi) [inline src/goto-instrument/contracts/contracts.cpp]: [nit] do we need to keep `cg` around? also, may be use a slightly more descriptive name for `g` and make it `const`?

```suggestion
  const auto call_graph = call_grapht::create_from_root_function(functions, root_function, true).get_directed_graph();
```
(would need to reformat ^)

File: src/goto-instrument/contracts/contracts.h
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/contracts.h]: [nit] this function isn't just "finding" static locals, it's also adding them to `assigns`. It's worth thinking about:
1. should we make this function "pure": only find static locals and return a `set` of those?
2. if not, would it make sense to move this function into the `assigns_clauset` class as a member function?
3. if not, should we rename this to `add_static_locals_to_assigns_set`?

**Review by SaswatPadhi — COMMENTED** [review]
Some minor comments:

File: src/goto-instrument/contracts/contracts.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/contracts.cpp]: the reason behind inlining is that in the function `f` we are instrumenting ,  a same function `g` can be called in several locations under different write sets. 
Without inlining, for each call site we would have to create a copy of g and instrument for the write set specific to this call site.
Inlining allows us to avoid having to duplicate g and instrument it specifically. 

As for inlining recursive functions, I'll implement a check to detect them and error out during inlining

File: src/goto-instrument/contracts/contracts.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/contracts.cpp]: we suppose function pointers have been removed/restricted before this instrumentation phase so that all possible target functions would appear explicitly in the call graph

Comment (tautschnig) [reply src/goto-instrument/contracts/contracts.cpp]: Hmm, we should really include a `PRECONDITION` in `call_graph.cpp`'s `forall_callsites` to replace `if(function_expr.id()==ID_symbol)` by `PRECONDITION_WITH_DIAGNOSTICS(function_expr.id() == ID_symbol, "call graph computation requires function pointer removal");`.

File: src/goto-instrument/contracts/contracts.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/contracts.cpp]: Mark this one as `override`

File: src/goto-instrument/contracts/contracts.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/contracts.cpp]: All of these need to be marked `override`.

File: src/goto-instrument/contracts/contracts.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/contracts.cpp]: Also to be marked `override`

File: src/goto-instrument/contracts/contracts.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/contracts.cpp]: Can be omitted.

File: src/goto-instrument/contracts/contracts.cpp
Comment (tautschnig) [inline src/goto-instrument/contracts/contracts.cpp]: These should be marked `virtual` in `util/message.h` and `override` in here.

File: src/goto-instrument/contracts/contracts.h
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/contracts.h]: I moved the method to the `assigns_clauset` class.

File: src/goto-instrument/contracts/contracts.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/contracts.cpp]: renamed as required

File: src/goto-instrument/contracts/contracts.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/contracts.cpp]: `PRECONDITION` added in `call_graph.cpp`. 


Should we also modify `goto_function_inline` to warn on unresolved function pointers ?

File: src/goto-instrument/contracts/contracts.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/contracts.cpp]: This method is indeed covered when running  `regression/contracts/assigns_enforce_statics_subfunctions` otherwise the test would not pass.

Comment (SaswatPadhi) [reply src/goto-instrument/contracts/contracts.cpp]: Thanks. Probably a false positive / stale result from Codecov then. Resolving this.

File: regression/contracts/function-calls-recursive-01/main.c
Comment (SaswatPadhi) [inline regression/contracts/function-calls-recursive-01/main.c]: [nit] missing newline

**Review by SaswatPadhi — APPROVED** [review]
LGTM 🎉

File: src/goto-instrument/contracts/contracts.cpp
Comment (tautschnig) [reply src/goto-instrument/contracts/contracts.cpp]: Yes, we should either warn or fail with a `PRECONDITION`.

**Comment by feliperodri** [comment]
@chrisr-diffblue @martin-cs @peterschrammel could you review this PR? It's missing code owner approval.

File: regression/contracts/function-calls-recursive-01/main.c
Comment (feliperodri) [reply regression/contracts/function-calls-recursive-01/main.c]: +1

File: src/goto-instrument/contracts/assigns.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/assigns.cpp:16]: [nit] this should go above `langapi` include(s).

File: src/goto-instrument/contracts/assigns.cpp
Comment (SaswatPadhi) [inline src/goto-instrument/contracts/assigns.cpp]: Redundant return

**Review by SaswatPadhi — COMMENTED** [review]
Couple of minor nitpicks but nothing that would block the merge.

File: src/goto-instrument/contracts/assigns.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/assigns.cpp:16]: fixed

File: src/goto-instrument/contracts/assigns.cpp
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/assigns.cpp]: fixed

**Review by tautschnig — APPROVED** [review]

File: src/goto-instrument/contracts/contracts.cpp
Comment (peterschrammel) [inline src/goto-instrument/contracts/contracts.cpp:29]: This is file is getting way to large. Consider refactoring/splitting in future.

**Review by peterschrammel — APPROVED** [review]
Nothing blocking, just a few remarks for future work:
That's quite a lot in a single commit. It could have been split into a series of commits:
- introduce decorator
- add new tests
- add static local detection
- other changes 1
- other changes 2
- other changes 3
- activate/modify tests (separately or with the corresponding commit that makes the test pass)

each of which with a 'why' description in the commit body to make the intent of the PR clearer and easier and faster to review.

**Pull request merged**
